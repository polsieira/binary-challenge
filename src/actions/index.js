export const checkIsLoading = (bool) => ({
  type: 'IS_LOADING',
  isLoading: bool
})

export const hasErrored = (message) => ({
  type: 'HAS_ERRORED',
  message
})

export const addExoplanets = (exoplanets) => ({
  type: 'ADD_EXOPLANETS',
  exoplanets
})

export const loginUser = (userInfo) => ({
  type: 'LOGIN_USER',
  userInfo
})

export const addFilters = (filters) => ({
  type: 'ADD_FILTERS',
  filters
})