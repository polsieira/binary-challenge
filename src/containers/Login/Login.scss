.form-sign-in {
  align-items: center;
  display: flex;
  flex-direction: column;
  justify-content: space-around;

  div {
    align-items: center;
    display: flex;
    flex-direction: column;
  }

  .form-label {

  }

  .form__input {
    height: 20px;
    margin-bottom: 20px;
    width: 100px;
  }

  .form__button {
    height: 20px;
    margin-bottom: 20px;
    width: 100px;
  }
}