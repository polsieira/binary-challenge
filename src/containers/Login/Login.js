import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import Button from '@material-ui/core/Button';
import TextField from '@material-ui/core/TextField';
import { hasErrored, loginUser, checkIsLoading } from '../../actions';
import './Login.scss';
import { Redirect } from 'react-router-dom';
import { bindActionCreators } from 'redux';
import { attemptLoginUser } from '../../apiCalls';



export class Login extends Component {
  constructor() {
    super();
    this.state = {
      email: '',
      password: '',
      isLoggedIn: false
    };
  }

  handleChange = event => {
    this.setState({
      [event.target.name]: event.target.value
    });
  };

  handleClick = async () => {
    const { hasErrored, loginUser } = this.props;
    const { email, password } = this.state;
    try {
      checkIsLoading(true);
      const userCheck = await attemptLoginUser({ email, password });
      checkIsLoading(false);
      this.setState({ isLoggedIn: true });
      loginUser({
        ...userCheck,
        isLoggedIn: true
      });
      hasErrored('');
    } catch ({ message }) {
      checkIsLoading(false);
      hasErrored(message);
    }
    this.clearInputs();
  };

  clearInputs = () => {
    const stateKeys = Object.keys(this.state);
    stateKeys.forEach(stateKey =>
      this.setState({
        [stateKey]: ''
      })
    );
  };

  handleSubmit = event => {
    event.preventDefault();
    this.handleClick();
  };



  render() {
    if (this.state.isLoggedIn) {
      return <Redirect to='/' />;
    }

    const { email, password } = this.state;
    const { error } = this.props;
    console.log(error)
    return (
      <div className='Login'>
        <h2 className={error ? 'login-message error-message' : 'login-message'}>{error ? error : 'Login in with email and password'}</h2>
        <TextField
          id='email'
          type='text'
          placeholder='your@email.com'
          className='form__input'
          name='email'
          value={email}
          onChange={e => this.handleChange(e)}
          label="Email"
          autoComplete="current-email"
          margin="normal"
          variant="filled"
          color="primary"
        />
        <TextField
          type="password"
          label="Password"
          id='password'
          className='form__input'
          name='password'
          placeholder='Must have at least 8 characters'
          value={password}
          onChange={e => this.handleChange(e)}
          autoComplete="current-password"
          margin="normal"
          color="secondary"
        />
        <Button className='form__button' primary="true" onClick={(event) => this.handleClick(event)}>
          Login
          </Button>
      </div >
    );
  }
}

export const mapStateToProps = ({ error, isLoading }) => ({
  error,
  isLoading
});

export const mapDispatchToProps = dispatch => {
  return bindActionCreators({
    hasErrored,
    loginUser,
    checkIsLoading,
  }, dispatch);
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Login);

Login.propTypes = {
  error: PropTypes.string,
  isLoading: PropTypes.bool,
  hasError: PropTypes.func
}