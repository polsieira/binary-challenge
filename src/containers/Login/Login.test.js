import React from 'react';
import { shallow } from 'enzyme';
import { Login, mapStateToProps, mapDispatchToProps } from './Login';
import { checkIsLoading, hasErrored, loginUser } from '../../actions';

jest.mock('../../apiCalls.js');

describe('Login', () => {
  describe('Login component', () => {
    let wrapper;

    beforeEach(() => {
      wrapper = shallow(<Login />);
    });

    it('should match the snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('should match the snapshot when there is an error', () => {
      wrapper = shallow(<Login error='error' />);
      expect(wrapper).toMatchSnapshot();
    })

    it('should update the email and password in state when handleChange is called', () => {
      const mockEventEmail = {
        target: {
          name: 'email',
          value: 'email@email.com'
        }
      };
      const mockEventPassword = {
        target: {
          name: 'password',
          value: 'password'
        }
      };

      const expectedEmail = 'email@email.com';
      const expectedPassword = 'password';

      wrapper.instance().handleChange(mockEventEmail);
      wrapper.instance().handleChange(mockEventPassword);

      expect(wrapper.state('email')).toEqual(expectedEmail);
      expect(wrapper.state('password')).toEqual(expectedPassword);
    });

    it('should clear all variables in state when clearInputs is invoked', () => {
      const expected = {
        email: '',
        password: '',
        isLoggedIn: ''
      };

      wrapper.instance().setState({
        email: 'email@email.com',
        password: 'password'
      });

      wrapper.instance().clearInputs();

      expect(wrapper.state()).toEqual(expected);
    });

    it('should call handleClick on click', () => {
      wrapper.instance().handleClick = jest.fn();
      wrapper.instance().forceUpdate();

      wrapper.find('.form__button').simulate('click');

      expect(wrapper.instance().handleClick).toHaveBeenCalled();
    });

  });

  describe('mapStateToProps', () => {
    it('should return an object with a movies array, an error message and isLoading property', () => {
      const mockState = {
        error: '',
        isLoading: false
      };

      const expected = {
        error: '',
        isLoading: false
      };

      const mappedProps = mapStateToProps(mockState);

      expect(mappedProps).toEqual(expected);
    });
  });

  describe('mapDispatchToProps', () => {

    const mockDispatch = jest.fn();

 

  });

});
