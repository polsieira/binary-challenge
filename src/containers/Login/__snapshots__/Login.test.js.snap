// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Login component should match the snapshot 1`] = `
<div
  className="Login"
>
  <h2
    className="login-message"
  >
    Login in with email and password
  </h2>
  <WithStyles(ForwardRef(TextField))
    autoComplete="current-email"
    className="form__input"
    color="primary"
    helperText="Incorrect email"
    id="email"
    label="Email"
    margin="normal"
    name="email"
    onChange={[Function]}
    placeholder="your@email.com"
    type="text"
    value=""
    variant="filled"
  />
  <WithStyles(ForwardRef(TextField))
    autoComplete="current-password"
    className="form__input"
    color="secondary"
    helperText="Incorrect password"
    id="password"
    label="Password"
    margin="normal"
    name="password"
    onChange={[Function]}
    placeholder="Must have at least 8 characters"
    type="password"
    value=""
  />
  <WithStyles(ForwardRef(Button))
    className="form__button"
    onClick={[Function]}
    primary="true"
  >
    Login
  </WithStyles(ForwardRef(Button))>
</div>
`;

exports[`Login Login component should match the snapshot when there is an error 1`] = `
<div
  className="Login"
>
  <h2
    className="login-message"
  >
    <Component />
  </h2>
  <WithStyles(ForwardRef(TextField))
    autoComplete="current-email"
    className="form__input form__input--error"
    color="primary"
    helperText="Incorrect email"
    id="email"
    label="Email"
    margin="normal"
    name="email"
    onChange={[Function]}
    placeholder="your@email.com"
    type="text"
    value=""
    variant="filled"
  />
  <WithStyles(ForwardRef(TextField))
    autoComplete="current-password"
    className="form__input form__input--error"
    color="secondary"
    helperText="Incorrect password"
    id="password"
    label="Password"
    margin="normal"
    name="password"
    onChange={[Function]}
    placeholder="Must have at least 8 characters"
    type="password"
    value=""
  />
  <WithStyles(ForwardRef(Button))
    className="form__button"
    onClick={[Function]}
    primary="true"
  >
    Login
  </WithStyles(ForwardRef(Button))>
</div>
`;
